book/src/proposals/tower-lockout-slashing.md

```diff
@@ -0,0 +1,95 @@
+# Tower Lockout Slashing
+
+This design describes how slashing is implemented in the solana
+cluster.  Slashing is designed as a punishment for participants in
+the cluster that violate the rules of the protocol.  A proof of the
+violation is submitted to the cluster and if the proof is accepted,
+the stake that is delegated to the node that violated a rule is
+slashed by some percentage.
+
+## Votes
+
+Validators submit votes which contain a list of slots.  To be
+accepted, the vote must contain the list of N slots currently in
+the VoteState program, along with new votes.
+
+The first `N - new votes`, must match the VoteState after running
```

Comment:
Can you provide some explanation/intuition here for why it's important for the first `N-new votes` to match the VoteState exactly?

Reply:
Also, we should probably mention what happens if these don't match. Is it a pre-req for the vote being accepted? If it doesn't happen is that vote simply dropped?

Reply:
@carllin the vote is just dropped.  So the validator created a vote with no reward but same slashing risk.  It must match the bank state to be accepted, which is why we can interpret the lockouts in those votes as real.

Reply:
So what happens if:

1)  a validator has voted for slots 1-40, so their local locktower looks like `9, 10..40`.
2) However only every other slot is landing in the bank, so this validator's vote account in the bank state is `2, 4, 8 .. 40`.

Now this validator's replay stage when it submits a vote will include his local locktower state, which doesn't match the bank, so his vote will get rejected every time? Do we need to change the voting behavior to submit votes based on the bank state of the fork you are voting on?

Reply:
@carllin yea. Basically validators have

1. A set of votes it has attempted to submit to the bank
2. A set of votes in the bank

Validators constrict the vote by taking what’s in the bank + missing votes.  If the network is censoring the validator, it needs to stop voting and alarm the operator.

Reply:
hmmm does that means once the first 20 votes land, you won't be able to submit any votes for the "gaps" in those 20 votes, which you can do today?

For instance in the example above, if only every other vote is landing, and the bank state is 2, 4, 8..40, I wouldn't be able to submit the missing votes 1, 3, 5, 7, etc. anymore

Reply:
@carllin how would there be gaps?  Validator votes have no gaps, so the bank won’t have gaps.

Reply:
Ah yeah, these votes carry all the history now, they're not individual votes anymore :D

Reply:
@aeyakovenko, one more point of clarification. In this design, we wouldn't be tracking all the validator's votes in an attempt to find any combination that would detect slashing? For instance if a validator voted:

`1, 2, 3`
`1, 2, 4`
`1, 7`

Does slashing then depend on order here when replaying these three votes? For instance, the first and second votes are  clearly slashable, but if the bank state plays the first vote then the third vote, the ending state will be `1, 7`. Then if second vote `1, 2, 4` was detected, that wouldn't be considered slashable when compared against `1, 7`.

Reply:
I think we only need two votes.

Reply:
hmm, you mean only track the last two votes? Doesn't seem like it solves the above issue?

Reply:
@carllin the fraud proof is composed of any two votes that show that the validator violated a lockout.

Reply:
@aeyakovenko, yeah that makes sense. The point of clarification I was trying to get in the above example was more so about who is responsible for doing the slashing detection, and how votes are being compared to different bank states to detect this slashing.

So in the above example, the validator has submitted these votes onto the network for three different forks:

`1, 2, 3`
`1, 2, 4`
`1, 7`

My question is

Who is doing the slashing detection for each of these forks?  Is it the leader that is trying to incorporate a vote to build on top of one of these forks? Is there a separate node like an archiver that is simultaneously keeping track of all `N` forks and also watching the network for any vote `V` that conflicts with the bank state on any of the `N` forks?

Reply:
@carllin we need gossip open, so ideally it’s everyone who is listening to gossip, and has copies of the ledger. Maybe archivers?

Reply:
@aeyakovenko ok, so is the proposal a validator pulls a vote off gossip, then compares it against every bank state it currently knows about to check for violations?

For example, if a validator knows about two different forks, with vote states:

1) `1, 2, 3`
2) `1, 7`

Then the validator sees the vote `1, 2, 4` and compares it against bank stats 1) and 2), detects a conflict with 1) and pushes a proof?

Reply:
It doesn’t need to do a bank state check.  It’s just looking at any two votes that overlap.

Reply:
@aeyakovenko votes don't _have_ to be sent over gossip, they can be sent to leaders directly. Especially in the case of malicious actors. 
Even if we required that to be the case, how do we keep track of multiple votes from the same peer? We only have 1 Vote Transaction per peer in gossip.

Reply:
@sagar-solana thats why I listed all the other ingress points.

Reply:
@sagar-solana gossip can be partitioned, and the sequential votes in gossip could still get slashed.

Reply:
@aeyakovenko, ah ok, so we are keeping track of all of a validator's votes, and every time a new one comes in, we compare it against all previous votes?

Reply:
If gossip is partitioned then how do we detect slashable votes? (sorry if you already answered this).
As for the sequancial votes, how many do we keep around?

Reply:
@carllin yep, but the slots in the Vote should make that a pretty fast check.  It just might be a fairly big data set.

Reply:
Maybe this should be part of gossip. We can run a slow lazy pull over the dataset.

Reply:
@aeyakovenko, so given two votes `A` and `B` to determine whether they are slashable, the protocol is:

1. Find whether `A` or `B` is older, meaning which one contains the highest voted slot. Let's say it's `B`

2. Find the slot `v_s` the vote that is least common ancestor of `A` and `B`. Let the first vote greater than `s` in  `A` and `B` be called `v_a` and `v_b` respectively

3. Apply `v_b` onto `A`. All votes in `A` `>= v_a` should be popped off. If not, these two votes are slashable.

Reply:
@aeyakovenko, you think the votes we store to detect slashing could be condensed into a trie based on shared prefixes (also store the signatures at each of the nodes)? I think that way for each validator, you store at most the number of slots/epoch worth of data to perform slashing, rather than number of votes/epoch of data.